File: Techniques/singleNaked.cpp
//
// Created by nathan on 29/02/24.
//

#include <vector>
#include <algorithm>
#include <iostream>
#include "singleNaked.h"
#include <unordered_set>
#include "../Tools/tools.h"

using namespace std;

void findSingleNaked(int grid[9][9]) {
    unordered_set<int> manquants(9);
    unordered_set<int> presents;

    for (int i = 0; i < 9; ++i) {
        manquants.clear();
        presents.clear();
        for (int j = 0; j < 9; ++j) {
            if (grid[i][j]) presents.insert(grid[i][j]);
        }

        for (int j = 0; j < 9; ++j) {
            if(!presents.contains(j))manquants.insert(j);
        }
        for (int j = 0; j < 9; ++j) { // <- parcours des cases de la ligne
            if (!grid[i][j]) {
                int indiceBloc = findBlocFromCoordinate(i, j);
                unordered_set<int> possibles = manquants;
                for (auto itr = possibles.begin();itr!= possibles.end();) {
                    if (i == 6) std::cout << possibles.size() << "\n";
                    if (isNumberInBloc(grid, indiceBloc, *itr) || isNumberInCol(grid, j, *itr))itr = possibles.erase(itr);
                    else++itr;
                }
                if (possibles.size() == 1) {
                    placeNumberInGrid(grid, i, j, *possibles.begin(), true);
                    std::cout << "Single Naked \n";
                }

            }
        }
    }
}

File: back/api/engine/singleNaked.go
package engine

import (
	"fmt"
)

// findSingleNaked identifies and places numbers when only one possibility exists in a row.
func FindSingleNaked(grid *Grid, numberPlaced *bool) []ResponseSolution {
	var row int8
	var solutions []ResponseSolution
	for row = 0; row < 9; row++ {
		manquants := make(map[int8]struct{}, 9)
		presents := make(map[int8]struct{}, 9)

		// Collect present numbers in the row
		for col := 0; col < 9; col++ {
			if grid.Grid[row][col].Value != 0 {
				presents[grid.Grid[row][col].Value] = struct{}{}
			}
		}

		var col int8
		for col = 1; col <= 9; col++ {
			if _, exists := presents[col]; !exists {
				manquants[col] = struct{}{}
			}
		}

		for col = 0; col < 9; col++ {
			if grid.Grid[row][col].Value == 0 {
				indiceBloc := FindBlocFromCoordinate(row, col)
				possibles := make(map[int8]struct{})
				if len(manquants) == 1 {
					for num := range manquants {
						grid.Grid[row][col].Value = num
						*numberPlaced = true
						fmt.Println("Single Naked")

						solutions = append(solutions, ResponseSolution{Solution{num, row, col, "Single Naked", "row"}, CreateCopy(grid)})

					}
				}
				// Copy manquants into possibles
				var num int8
				for num = range manquants {
					possibles[num] = struct{}{}
				}

				// Remove numbers that are already in the block or column
				for num := range possibles {

					if IsNumberInCol(grid, col, num) {
						delete(possibles, num)
					}
				}
				if len(possibles) == 1 {
					for num := range possibles {
						grid.Grid[row][col].Value = num
						*numberPlaced = true
						fmt.Println("Single Naked")

						solutions = append(solutions, ResponseSolution{Solution{num, row, col, "Single Naked", "col"}, CreateCopy(grid)})

					}
				} else {
					for num := range possibles {

						if IsNumberInBloc(grid, indiceBloc, num) || IsNumberInCol(grid, col, num) {
							delete(possibles, num)
						}
					}

					if len(possibles) == 1 {
						for num := range possibles {
							grid.Grid[row][col].Value = num
							*numberPlaced = true
							fmt.Println("Single Naked")

							solutions = append(solutions, ResponseSolution{Solution{num, row, col, "Single Naked", ""}, CreateCopy(grid)})

						}
					}
				}

			}
		}
	}
	return solutions
}

// Helper functions (isNumberInBlc, isNumberInCol, findBlocFromCoordinate, and placeNumberInGrid) need to be implemented.
